GH-3135 : Adding new `ListenerContainerRegistry.getListenerContainersMatchingId(Predicate)` for easier usage.

JooHyukKim opened the pull request:
Resolves #3135

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/config/KafkaListenerEndpointRegistry.java:
```diff
@@ -115,6 +118,25 @@ public MessageListenerContainer getListenerContainer(String id) {
+			.stream()
+			.filter(entry -> idMatcher.test(entry.getKey()))
+			.map(Map.Entry::getValue)
+			.collect(Collectors.toList());
```
There is just `toList()` API.

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/config/KafkaListenerEndpointRegistry.java:
```diff
@@ -115,6 +118,25 @@ public MessageListenerContainer getListenerContainer(String id) {
+	 * @param idMatcher the predicate to match container id with
+	 * @return the containers or empty {@link Collection} if no container with that id exists
+	 * @see #getListenerContainerIds()
+	 * @see #getListenerContainer(String)
```
`@since 3.2`

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/listener/ListenerContainerRegistry.java:
```diff
@@ -41,6 +42,16 @@ public interface ListenerContainerRegistry {
+	 * @return the containers or empty {@link Collection} if no container with that id exists
+	 * @see #getListenerContainerIds()
+	 * @see #getListenerContainer(String)
+	 */
```
`@since 3.2`

artembilan commented on spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java:
```diff
@@ -47,4 +63,77 @@ void unregister() {
+        // When
+        Collection<MessageListenerContainer> listeners = registry.getListenerContainersMatching(s -> true);
+        // Then
+        assertThrows(UnsupportedOperationException.class, () -> listeners.add(mock(MessageListenerContainer.class)));
```
Why don't use `assertThatExceptionOfType(UnsupportedOperationException.class)` instead?

It also probably going to fail on Checkstyle since we have a limited set of allowed `static` method imports.

JooHyukKim replied:
Right, looking now there is no pre-existing usage of `assertThatExceptionOfType()`

artembilan commented on spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java:
```diff
@@ -47,4 +63,77 @@ void unregister() {
+		names.forEach(name -> registerListenerWithId(registry, name));
+	}
+
+	private void registerListenerWithId(KafkaListenerEndpointRegistry registry, String id) {
```
Looks like these both methods can be `static`.

artembilan reviewed (changes requested):
Please, add your name to the `@author` list of all the affected classes.

Also consider to mention this new feature in the `whats-new.adoc` and in the `kafkalistener-lifecycle.adoc`.

Thanks

JooHyukKim commented:
I've updated the author list and added documentation for the new feature in `whats-new.adoc` and `kafkalistener-lifecycle.adoc`.

I noticed there seems to be limited documentation on `KafkaListenerEndpointRegistry`, particularly methods like `getListenerContainers()` and `getListenerContainer(String)`. I can
 address this in a future PR to enhance docs further. WDYT, @artembilan?

artembilan commented:
Sure! You can add those couple sentences just in this PR. We have a release on Monday, so it will be reviewed immediately .
Thanks

JooHyukKim commented:
Added some more doc as discussed 👍🏼 cc. @artembilan

artembilan commented on spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:
```diff
@@ -83,3 +83,13 @@ See xref:kafka/seek.adoc#seek[Seek API Docs] for more details.
+== Retrieve MessageListenerContainers with by matching a Predicate
+
+Starting with version 3.2, the `ListenerContainerRegistry` introduces the `getListenerContainersMatching` method.
+This method filters `MessageListenerContainer` instances based on a `Predicate<String>` applied to their IDs.
```
After your great docs I wonder if we could extend this new feature to something like this `BiPredicate<String, MessageListenerContainer>`. This way we may filter desired containers not only by their ids, but via some other option form the container itself.

WDYT?

JooHyukKim replied:
Makes sense 👍🏼
I wonder if we should have both or just `BiPredicate<String, MessageListenerContainer>` 🤔
Extended feature sounds very powerful, but might be too verbose in someway, knowing that `id` is the key used to access the `MessageContainerListener`.

JooHyukKim replied:
Since the `String, MessageListenerContainer` pair in `BiPredicate` technically represents the key-value pair stored in the registry, how about having both?

JooHyukKim replied:
Should we do a separate PR or do everything in this one?

artembilan replied:
This one is OK.
Thanks

JooHyukKim replied:
Alright, hopefully it would probably take an hour or two

artembilan commented on spring-kafka/src/main/java/org/springframework/kafka/listener/ListenerContainerRegistry.java:
```diff
@@ -41,6 +44,29 @@ public interface ListenerContainerRegistry {
+	 * @see #getListenerContainerIds()
+	 * @see #getListenerContainersMatching(Predicate)
+	 */
+	Collection<MessageListenerContainer> getListenerContainersMatching(BiPredicate<String, MessageListenerContainer> idAndContainerMatcher);
```
Let's see if you can follow the rule no more than 120 symbols per line!

artembilan commented on spring-kafka-docs/src/main/antora/modules/ROOT/pages/whats-new.adoc:
```diff
@@ -83,3 +83,26 @@ See xref:kafka/seek.adoc#seek[Seek API Docs] for more details.
+[[get-listener-containers-matching]]
+== Enhanced Retrieval of MessageListenerContainers
+
+Introduced in version 3.2, the `ListenerContainerRegistry` provides advanced methods for dynamically filtering `MessageListenerContainer` instances, enhancing management capabilities.
```
This file is fully specific to particular version we mention in the beginning of the file.
More over the content of the file is just a list of topics with links to details.
You do show enough in the `kafkalistener-lifecycle.adoc`, but that all must not be duplicated in `What's New` file.

JooHyukKim replied:
True, thank you for pointing out!

artembilan commented on spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java:
```diff
@@ -47,4 +60,123 @@ void unregister() {
+		given(container.isRunning()).willReturn(true);
+		given(factory.createListenerContainer(endpoint)).willReturn(container);
+		registry.registerListenerContainer(endpoint, factory);
+	}
```
Every member (including the last method of the last class) in the class must be surrounded with blank lines for much better code readability.

artembilan reviewed (changes requested):
The API is great: only some cosmetic concerns in regards of our code style rules.
Thanks
